src/app/cli/src/init/mina_run.ml

```diff
@@ -21,6 +21,123 @@ let snark_pool_list t =
   |> Network_pool.Snark_pool.Resource_pool.snark_pool_json
   |> Yojson.Safe.to_string
 
+let start_auto_hardfork_config_generation ~logger mina =
+  let open Deferred.Let_syntax in
+  let config = Mina_lib.config mina in
+  let runtime_config = config.precomputed_values.runtime_config in
+  let consensus_constants = config.precomputed_values.consensus_constants in
+  match config.hardfork_handling with
+  | Cli_lib.Arg_type.Hardfork_handling.Keep_running ->
+      Deferred.unit
+  | Cli_lib.Arg_type.Hardfork_handling.Migrate_exit -> (
+      match
+        ( Runtime_config.slot_chain_end runtime_config
+        , Runtime_config.slot_tx_end runtime_config
+        , Mina_lib.scheduled_hard_fork_genesis_slot mina )
+      with
+      | Some slot_chain_end, Some slot_tx_end, Some _scheduled_genesis -> (
+          (* Calculate when slot_chain_end begins *)
+          let chain_end_time =
+            Consensus.Data.Consensus_time.(
+              start_time ~constants:consensus_constants
+                (of_global_slot ~constants:consensus_constants slot_chain_end))
+          in
+          let chain_end_time_str =
+            Block_time.to_time_exn chain_end_time
+            |> Time.to_string_iso8601_basic ~zone:Time.Zone.utc
+          in
+          [%log debug]
+            "Spawning background thread to wait for hardfork config \
+             generation. Will wait until slot_chain_end: slot %s at %s"
+            (Mina_numbers.Global_slot_since_hard_fork.to_string slot_chain_end)
+            chain_end_time_str ;
+          (* Wait until slot_chain_end time *)
+          let%bind () =
+            let span =
+              Block_time.diff chain_end_time
+                (Block_time.now config.time_controller)
+            in
+            after (Block_time.Span.to_time_span span)
+          in
+          [%log debug]
+            "Reached slot_chain_end time, now waiting for best tip to reach \
+             slot_tx_end: %s"
+            (Mina_numbers.Global_slot_since_hard_fork.to_string slot_tx_end) ;
+          (* Now poll until best tip >= slot_tx_end *)
+          let rec wait_for_best_tip () =
```

Comment:
suggest wrap this inside `Deferred.repeat_until_finished`, maybe also lift this to top level